### PR TITLE
Assert not same executor when completing future

### PR DESCRIPTION
**Pull request by henningandersen**
A common deadlock pattern is waiting and completing a future on the same executor. This only works until the executor is fully depleted of threads. Now assert that waiting for a future to be completed and the completion happens on different executors.

Introduced UnsafePlainActionFuture, used in all offending places, allowing those to be tackled independently.

### CODE REVIEW

**Comment by henningandersen** [comment]
Opening as draft for now to see how CI behaves.

File: server/src/main/java/org/elasticsearch/common/util/concurrent/EsExecutors.java
Comment (DaveCTurner) [inline server/src/main/java/org/elasticsearch/common/util/concurrent/EsExecutors.java:272]: Probably want to return `false` here rather than failing with an unhelpful message?

File: server/src/main/java/org/elasticsearch/common/util/concurrent/EsExecutors.java
Comment (DaveCTurner) [inline server/src/main/java/org/elasticsearch/common/util/concurrent/EsExecutors.java:280]:
```suggestion
        assert Assertions.ENABLED;
```

This is a no-op: if assertions are disabled, we don't even run this check :)

**Review by DaveCTurner — COMMENTED** [review]
Nice idea. I think we can make this even stronger tho, we can still deadlock if we have a loop of executors all waiting on each other to complete. Can we declare up-front all pairs of executors that can be in a block/complete relationship and verify that this permits no such cycles?

File: server/src/main/java/org/elasticsearch/common/util/concurrent/EsExecutors.java
Comment (henningandersen) [reply server/src/main/java/org/elasticsearch/common/util/concurrent/EsExecutors.java:272]: This was sort of on purpose in that, the method should only be used in situations where you are clearly comparing different threads. It seems that if you use this for the same thread your assertion is not written correctly. Hence more of an assertion on the assertion. Not too tied to this though.

File: server/src/main/java/org/elasticsearch/common/util/concurrent/EsExecutors.java
Comment (henningandersen) [reply server/src/main/java/org/elasticsearch/common/util/concurrent/EsExecutors.java:280]: oh, I fixed that in one place already, thanks.

File: server/src/main/java/org/elasticsearch/common/util/concurrent/EsExecutors.java
Comment (DaveCTurner) [reply server/src/main/java/org/elasticsearch/common/util/concurrent/EsExecutors.java:272]: Oh I see, yeah, we can't be completing on the waiting thread because that thread is waiting...

(deserves a comment to avoid redoing that thinking tho)

File: server/src/main/java/org/elasticsearch/common/util/concurrent/EsExecutors.java
Comment (kingherc) [inline server/src/main/java/org/elasticsearch/common/util/concurrent/EsExecutors.java:281]: nit A comment on why `name.length() - 2` would be helpful. Why is it like that? I understand that probably there's another `]` you'd like to skip? But I don't remember how the thread name is formatted out of the top of my head. Reading

```
    public static String threadName(final String nodeName, final String namePrefix) {
        // TODO missing node names should only be allowed in tests
        return "elasticsearch" + (nodeName.isEmpty() ? "" : "[") + nodeName + (nodeName.isEmpty() ? "" : "]") + "[" + namePrefix + "]";
    }
```

and assuming namePrefix is the executor name, I'm not sure where the other `]` would be.

**Review by kingherc — COMMENTED** [review]
Nice idea!

**Comment by henningandersen** [comment]
> Nice idea. I think we can make this even stronger tho, we can still deadlock if we have a loop of executors all waiting on each other to complete. Can we declare up-front all pairs of executors that can be in a block/complete relationship and verify that this permits no such cycles?

Sounds like a good idea. For now, I'd like to ensure this can be made to pass CI at all, I think there could be many cases of just this first assertion failing. I might defer your suggestion to a follow-up.

File: server/src/main/java/org/elasticsearch/common/util/concurrent/EsExecutors.java
Comment (henningandersen) [reply server/src/main/java/org/elasticsearch/common/util/concurrent/EsExecutors.java:270]: I had that first, but since it no longer asserts that the executors are different but rather returns whether they are, I renamed it to avoid confusion.

File: server/src/main/java/org/elasticsearch/common/util/concurrent/EsExecutors.java
Comment (henningandersen) [reply server/src/main/java/org/elasticsearch/common/util/concurrent/EsExecutors.java:283]: The thread naming is something like `XXXX[executor-name][T#NNN]`. We need to skip the running number part. The index passed into `lastIndexOf` is inclusive.

I added a short comment on this.

Comment (fcofdez) [reply server/src/main/java/org/elasticsearch/common/util/concurrent/EsExecutors.java:283]: Maybe we could use a regex here? maybe that's clearer? (famous last words)

**Review by fcofdez — COMMENTED** [review]
This looks good

File: server/src/main/java/org/elasticsearch/common/util/concurrent/EsExecutors.java
Comment (DaveCTurner) [reply server/src/main/java/org/elasticsearch/common/util/concurrent/EsExecutors.java:283]: This time there's even _two_ obligatory XCKDs:
- https://xkcd.com/208/
- https://xkcd.com/1171/

File: server/src/main/java/org/elasticsearch/index/engine/Engine.java
Comment (henningandersen) [inline server/src/main/java/org/elasticsearch/index/engine/Engine.java:1973]: These changes will not go in as is, but wanted to get more build failures out.

File: server/src/main/java/org/elasticsearch/common/util/concurrent/EsExecutors.java
Comment (henningandersen) [reply server/src/main/java/org/elasticsearch/common/util/concurrent/EsExecutors.java:283]: I prefer the manual parsing done here, seems simplest.

File: server/src/main/java/org/elasticsearch/action/support/UnsafePlainActionFuture.java
Comment (henningandersen) [inline server/src/main/java/org/elasticsearch/action/support/UnsafePlainActionFuture.java:23]: I opted to add this unsafe variant that is then used in all places where I found conflicts, since fixing them all would make this PR size unmanageable. Using this we can now fix them one by one, spreading out the load too.

**Comment by henningandersen** [comment]
This now has a clean build so marking ready for review. I'll do more CI work to scrape up more occurrences before merge anyway (which can result in more unsafe future usages).

File: server/src/main/java/org/elasticsearch/action/support/UnsafePlainActionFuture.java
Comment (DaveCTurner) [reply server/src/main/java/org/elasticsearch/action/support/UnsafePlainActionFuture.java:23]: :+1: can we mark this as `@Deprecated(forRemoval = true)` so that IDEs highlight its usages?

File: server/src/main/java/org/elasticsearch/client/internal/support/AbstractClient.java
Comment (DaveCTurner) [inline server/src/main/java/org/elasticsearch/client/internal/support/AbstractClient.java:416]: Yikes. I suspect we can/should move all usages of this into tests, and do the ref-counting (and asyncification) properly in prod code. But I see that's not a small change. Ok for now...

**Review by DaveCTurner — APPROVED** [review]
Great stuff, I like it.

File: server/src/main/java/org/elasticsearch/action/support/UnsafePlainActionFuture.java
Comment (kingherc) [reply server/src/main/java/org/elasticsearch/action/support/UnsafePlainActionFuture.java:23]: Do we need a task/issue/ticket on fixing them one-by-one?

**Review by kingherc — APPROVED** [review]
LGTM as long as CI is happy

**Review by tlrx — APPROVED** [review]
LGTM, nice idea!

File: server/src/main/java/org/elasticsearch/action/support/UnsafePlainActionFuture.java
Comment (henningandersen) [reply server/src/main/java/org/elasticsearch/action/support/UnsafePlainActionFuture.java:23]: I think one by one is too granular. I'll probably create area level ones (perhaps a bit more granular, for instance the one in `AbstractClient` needs a separate one).

File: server/src/main/java/org/elasticsearch/client/internal/support/AbstractClient.java
Comment (henningandersen) [reply server/src/main/java/org/elasticsearch/client/internal/support/AbstractClient.java:416]: Yeah, I also pondered on this for a short while, but decided to defer this issue for now. I think I agree to move to all prod client interactions being async now. If it is too big we would need to stop notifying on generic thread pool (one more pool maybe, hopefully as an interim step). We can discuss more when we tackle it.

**Pull request merged**
